wgengine/tsdns/dns.go

```diff
@@ -0,0 +1,254 @@
+	offset += 1
+
+	if len(in) < offset+4 {
+		return ErrIncomplete
+	}
+
+	typ := binary.BigEndian.Uint16(in[offset : offset+2])
+	class := binary.BigEndian.Uint16(in[offset+2 : offset+4])
+	if typ != typeA || class != classIN {
+		return ErrUnknownTypeClass
+	}
+
+	return nil
+}
+
+// writeReply assembles a DNS reply packet from a message struct.
+// It returns the number of bytes written or an error.
+//
+// The structure of a DNS reply is as follows:
+// (c for copied from query, x for don't care)
+//
+//       0123 4567 89ab cdef 0123 4567 89ab cdef
+//      ┌───────────────────┬───────────────────┐
+//      │   Transaction ID  │       Flags       │
+//   0  │         c         │1000 0xxc xxcx xxxx│
+//      ├───────────────────┼───────────────────┤
+//      │   Question count  │  Answer RR count  │
+//   4  │         1         │         1         │
+//      ├───────────────────┼───────────────────┤
+//      │Authority RR count │Additional RR count│
+//   8  │         0         │         0         │
+//      ├───────────────────┴───────────────────┤
+//      │              Question 0               │
+//      ├───────────────────────────────────────┤
+//      │                 Name                  │
+//  12  │              c (n bytes)              │
+//      │                                       │
+//      │              name.ipn.dev             │
+//      │                   ↕                   │
+//      │      \x03name\x03ipn\x03dev\x00       │
+//      ├───────────────────┬───────────────────┤
+//      │       Type        │       Class       │
+// 12+n │     0x01 (A)      │     0x01 (IN)     │
+//      ├───────────────────┴───────────────────┤
+//      │               Answer 0                │
+//      ├───────────────────────────────────────┤
+//      │                 Name                  │
+// 16+n │              c (n bytes)              │
+//      │                                       │
+//      │              name.ipn.dev             │
+//      │                   ↕                   │
+//      │      \x03name\x03ipn\x03dev\x00       │
+//      ├───────────────────┬───────────────────┤
+//      │       Type        │       Class       │
+// 16+2n│     0x01 (A)      │     0x01 (IN)     │
+//      ├───────────────────┴───────────────────┤
+//      │                  TTL                  │
+// 20+2n│                 3600                  │
+//      ├───────────────────┬───────────────────┤
+//      │       Length      │         IP        │
+// 24+2n│        0x04       │      100.xxx      │
+//      ├───────────────────┼───────────────────┤
+//      │         IP        │///////////////////│
+// 28+2n│      yyy.zzz      │///////////////////│
+//      └───────────────────┴───────────────────┘
+func writeReply(m *message, out []byte) (int, error) {
+	if len(out) < minResponseSize+len(m.Question.Name) {
+		return 0, ErrSmallBuffer
+	}
+
+	binary.BigEndian.PutUint16(out[0:2], m.Header.TransactionID)
```

Comment:
I tend to define a package global when working exclusively with one endian:

```
var bin = binary.BigEndian
```

to make these lines a little shorter

Reply:
https://golang.org/pkg/encoding/binary/#Write (and Read) might also help.

Reply:
I will give that a try if the dnsmessage packet doesn't meet our needs.